src/beamlime/constructors/inspectors.py

```diff
@@ -109,33 +121,47 @@ def get_product_spec(callable_obj: Callable[..., Product]) -> ProductSpec:
         return ProductSpec(product)
 
 
-def extract_underlying_dep_type(tp: Any) -> Any:
-    from typing import Union, get_args, get_origin
-
-    if get_origin(tp) == Union:
-        if len((args := get_args(tp))) == 2 and type(None) in args:
-            # Optional
-            return args[0] if isinstance(None, args[1]) else args[1]
-        else:
-            raise NotImplementedError(
-                "Union annotation for dependencies is not supported."
-            )
-    else:
-        return tp
```

Comment:
Why did you move it?

Reply:
Just to specify the scope of this function more in detail and distinguish it from `extract_underlying_type`.
This one is only for argument (dependency) type-checks.

And `extract_underlying_type` is used both for argument and returned type handling.